### DISCUSSION TITLE
Understanding AST and visitToken() function

### DISCUSSION THREAD

**Discussion by TheSuperEngineer** [checkstyle/checkstyle#12907] [Q&A]
### Here is a conversation that I had with chatGPT:

I was trying to understand what is the purpose of AST here:
![image](https://user-images.githubusercontent.com/90137881/227713932-3b4aa88f-2184-4ae7-8f35-89b2057e89e2.png)

An arbitrary source code that I have doubt about (kindly read below):

```java
import com.puppycrawl.tools.checkstyle.api.DetailAST;

public class ExampleCheck {

    public void traverseAST(DetailAST ast) {
        DetailAST node = ast.getFirstChild();
        while (node != null) {
            // Process the current node and print its type and text content
            System.out.println("Node type: " + node.getType() + ", text: " + node.getText());

            // Move to the next sibling node
            node = node.getNextSibling();
        }
    }

    public static void main(String[] args) {
        String source = "public class Example {\n" +
                "    public static void main(String[] args) {\n" +
                "        int x = 1;\n" +
                "        if (x == 1) {\n" +
                "            System.out.println(\"x is 1\");\n" +
                "        } else {\n" +
                "            System.out.println(\"x is not 1\");\n" +
                "        }\n" +
                "    }\n" +
                "}";
        ExampleCheck check = new ExampleCheck();
        DetailAST ast = DetailASTUtil.parse(source);
        check.traverseAST(ast);
    }

}

```
#### Is the above java code a smaller representation of what happens in checkstyle codebase when working on source codes? Is my line of thought correct below.

When a user runs the checkstyle application in the terminal using the command ``java -jar "checkstyle-10.7.0-all.jar" -c config.xml test.java ``, the source code is taken from ``test.java`` in the format of a large ``String``.  Since, parsing on just the source code is not beneficial as making checks, formatting etc would be difficult, we convert the ``String`` (that is ``test.java``) into an AST using ``DetailAST ast = DetailASTUtil.parse(source);`` where ``ast`` holds the whole AST of ``test.java``.

Since, working with the AST is much more simpler and beneficial (as it gives more control over the code), we use the AST to make the checks etc and report violation. 

This is where ``visitToken()`` comes into the picture. ``visitToken(DefaultAST)`` takes the ``ast`` (which is the AST of ``test.java``) and helps perform all kinds of operations like going to the ``nextSibling``, going to the ``parentNode``, going to the leaf nodes etc.
This automatically requires for ``visitToken()`` to be present in every file of the source code of checkstyle that is involved in making checks on the file sent by the user which is ``test.java``.

Information that chatGPT gave me:
![image](https://user-images.githubusercontent.com/90137881/227714867-3a4e18f9-e913-43db-acba-9ec70700ab70.png)
![image](https://user-images.githubusercontent.com/90137881/227714906-b90db724-eae5-481a-a780-789f7dc7dd02.png)

**Comment by romani** [comment]
kind of yes, you need to read https://checkstyle.org/writingchecks.html#Understanding_the_visitor_pattern

**Reply by nrmancuso** [reply]
@shubh220922 to add to what @romani mentioned above, our abstract syntax tree (AST) and how we visit certain nodes in the tree (visittoken method) for our checks are separate concepts; I would suggest opening a new discussion about our AST and how that is created, if you would like to know more about that topic. Just like in issues and PRs, `and` in titles are usually a red flag that we can separate ideas to make it easier to understand and reason about :)

**Reply by TheSuperEngineer** [reply]
@nrmancuso noted.

I will spend few hours studying both of the concepts before i break into two different discussions for this.

**Reply by nrmancuso** [reply]
@shubh220922 if you want to play with AST idea on a smaller scale, you can clone my project at https://github.com/nrmancuso/parseva-math :) I made this as a proof of concept for my last GSOC project (ANTLR2 -> ANTLR4 update).

**Reply by TheSuperEngineer** [reply]
linked to https://github.com/checkstyle/checkstyle/discussions/12912

**Comment by rnveach** [comment]
> Is the above java code a smaller representation of what happens in checkstyle codebase when working on source codes? 

No. Your code only goes from sibling to sibling like a production line or 1 dimensional array. It has no depth to it. If you are familar with linked lists, then you need to think of Checkstyle's AST as a 2 dimensional linked list. If your not familar with linked lists, then I would suggest looking it up. We are virtually the same as a linked list.

Your code is missing usages of `getFirstChild` and `getParent` to travel through the 2nd dimension.

If you want to simulate what CS does for "walking" through the nodes, one by one, then look at https://github.com/checkstyle/checkstyle/blob/6a36be871dd183304f76f73fa27045b59f0ea5c4/src/main/java/com/puppycrawl/tools/checkstyle/TreeWalker.java#L403-L414 . Once you become familiar with this pattern, you start to see it in other areas of the code dealing with "walking".
Example: https://github.com/checkstyle/checkstyle/blob/c9190ed2fbb919beafeb992bbce98912c6fa8717/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/AbstractJavadocCheck.java#L362-L386

> This automatically requires for visitToken() to be present in every file of the source code of checkstyle that is involved in making checks on the file sent by the user which is test.java.

Technically no. `visitToken` is only present in the source code when dealing with the AST which is derived Java files. We have `AbstractFileSet` which do not have a `visit` method because it deals with the String source and not an AST representation. We also have filters which deal more with violation objects. Etc...

Also, you are not required to use the `visitToken` method even for AST checks. By default the method does nothing and does not need to be implemented. You can implement a check using only `beginTree`/`finishTree`/`leaveToken` depending on what kind of check you want. I do believe we have at least 1 check in our codebase that does not use `visitToken` and does something with one of the others.

**Reply by TheSuperEngineer** [reply]
@rnveach yes i'm totally familiar with linked lists (I'm getting to know about 2-D linkedlist for the first time but, the concept seems normal), graphs, tree etc and thankyou for the explanation. I'm diving deeper into the topic because very new things like the idea of visitor pattern etc has come across for me.
